packages/template-retail-react-app/app/components/dynamic-image/index.jsx

```diff
@@ -1,37 +1,107 @@
+const DynamicImage = ({src, widths, densities, imageProps, as, ...rest}) => {
+    const Component = as ? as : Img
     const theme = useTheme()
 
-    const responsiveImageProps = useMemo(
-        () => getResponsiveImageAttributes({src, widths, breakpoints: theme.breakpoints}),
-        [src, widths, theme.breakpoints]
-    )
+    const [responsiveImageProps, l, effectiveImageProps, responsiveLinks] = useMemo(() => {
+        const responsiveImageProps = getResponsivePictureAttributes({
+            src,
+            widths,
+            densities,
+            breakpoints: theme.breakpoints
+        })
+        const l = responsiveImageProps.sources.length
+        const effectiveImageProps = getImageAttributes(imageProps)
+        const fetchPriority = effectiveImageProps.fetchPriority
+        const responsiveLinks =
+            !responsiveImageProps.links.length && fetchPriority === 'high'
+                ? [
+                      getImageLinkAttributes({
+                          ...effectiveImageProps,
+                          fetchPriority, // React <18 vs. >=19 issue
+                          src: responsiveImageProps.src
+                      })
+                  ]
+                : responsiveImageProps.links.reduce((acc, link) => {
+                      const linkProps = getImageLinkAttributes({
+                          ...effectiveImageProps,
+                          ...link,
+                          fetchPriority, // React <18 vs. >=19 issue
+                          src: responsiveImageProps.src
+                      })
+                      if (linkProps) {
+                          acc.push(linkProps)
+                      }
+                      return acc
+                  }, [])
+        return [responsiveImageProps, l, effectiveImageProps, responsiveLinks]
+    }, [src, widths, densities, theme.breakpoints])
 
     return (
         <Box {...rest}>
-            <Component {...responsiveImageProps} {...imageProps} />
+            {l > 0 ? (
+                <picture>
+                    {responsiveImageProps.sources.map(({srcSet, sizes, media, src}, idx) => {
+                        if (idx < l - 1) {
+                            return <source key={idx} media={media} sizes={sizes} srcSet={srcSet} />
+                        }
+                        return (
+                            <Component
+                                key={idx}
+                                {...effectiveImageProps}
+                                sizes={sizes}
+                                srcSet={srcSet}
+                                src={responsiveImageProps.src}
+                            />
+                        )
+                    })}
+                </picture>
+            ) : (
+                <Component {...effectiveImageProps} src={responsiveImageProps.src} />
+            )}
+
+            {isServer() && responsiveLinks.length > 0 && (
+                <Helmet>
```

Comment:
Dumb question, why rendering this only server side and not both?

Reply:
Because the effect is completely insignificant during CSR-mode. The impact of preload links is only relevant for SSR'ed pages and once the app bootstrapped and CSR navigation kicks in, it'd be more overhead than benefit. Things would be different if we switched to an MPA mode.